### PR TITLE
[BEAM-8575] Test a customized window fn work as expected

### PR DESCRIPTION
**Pull request by liumomo315**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by liumomo315** [comment]
Run Python Dataflow ValidatesRunner

**Comment by liumomo315** [comment]
R: @lcwik

File: sdks/python/apache_beam/transforms/window_test.py [hunk 1]
```diff
@@ -252,6 +276,50 @@ def test_timestamped_with_combiners(self):
...
+
+      assert_that(result, equal_to([('key', [0, 1, 2]),
+                                    ('key', [3, 4]),
+                                    ('key', [5]),
+                                    ('key', [6])]))
+
+  @attr('ValidatesRunner')
+  def test_windows_idempotency(self):
+    with TestPipeline() as p:
+      result = (p | Create([(x, x * 2) for x in range(5)])
```
Comment (lukecwik) [inline sdks/python/apache_beam/transforms/window_test.py]: You should create timestamped values in Create instead of down below.

File: sdks/python/apache_beam/transforms/window_test.py [hunk 2]
```diff
@@ -252,6 +276,50 @@ def test_timestamped_with_combiners(self):
...
+                | GroupByKey())
+
+      assert_that(result, equal_to([('key', [0, 1]),
+                                    ('key', [2, 3]),
+                                    ('key', [4])]))
+
+  @attr('ValidatesRunner')
+  def test_windows_gbk_idempotency(self):
+    with TestPipeline() as p:
+      result = (p | Create([(x, x * 2) for x in range(5)])
```
Comment (lukecwik) [inline sdks/python/apache_beam/transforms/window_test.py]: Ditto for timestamped values being done in Create.

File: sdks/python/apache_beam/transforms/window_test.py (hunk 1 above)
Comment (liumomo315) [reply sdks/python/apache_beam/transforms/window_test.py]: I tried that but it failed.

Dug this issue more and found this is a known issue with Python timestamped value, see https://github.com/apache/beam/pull/8206. In short, to have the downstream ptransforms treat an timestamped element as an element with a timestamp instead of a timestamped value object, there must be a DoFn (could be a simple Map(lamda x: x)) applied to the timestamped elements. (This should be fixed, but probably not in this PR....)

I found this test class already have a helper fun timestamped_key_values, so I reuse it in the new tests.

File: sdks/python/apache_beam/transforms/window_test.py (hunk 2 above)
Comment (liumomo315) [reply sdks/python/apache_beam/transforms/window_test.py]: Same for this one:)

File: sdks/python/apache_beam/transforms/window_test.py [hunk 3]
```diff
@@ -65,6 +76,44 @@ def process(self, element, window=core.DoFn.WindowParam):
...
+      return [IntervalWindow(0, self.first_window_end)]
+    elif timestamp < self.second_window_end:
+      return [IntervalWindow(self.first_window_end, self.second_window_end)]
+    else:
+      return [IntervalWindow(self.second_window_end, timestamp)]
+
+  def get_window_coder(self):
+    return IntervalWindowCoder()
+
+  def to_runner_api_parameter(self, context):
```
Comment (liumomo315) [inline sdks/python/apache_beam/transforms/window_test.py]: This custom window fn is added only for a test, to test a non-standard window fn. To fully implement it, I need to also implement to_runner_api_parameter and from_runner_api_parameter. Those require me to also change the standard_window_fns.proto and common_urns.py. Are these changes necessary? If not, is there a better way to do this? Thanks! @lukecwik @robertwb

**Comment by liumomo315** [comment]
> Welcome to Beam, all non-trivial PRs are meant to be accompanied with a JIRA in the title as per the [contribution guide](https://beam.apache.org/contribute/#share-your-intent).
> 
Thanks for the pointer. Added JIRA:)

> Also, please address the lint issue:
> 
> ```
> Task :sdks:python:test-suites:tox:py37:lintPy37
> 15:45:38 ************* Module apache_beam.transforms.window_test
> 15:45:38 apache_beam/transforms/window_test.py:283:48: E0110: Abstract class 'TestCustomWindows' with abstract methods instantiated (abstract-class-instantiated)
> ```
Done fully implementing it, but have a question on whether I could make it simpler without changing other files, in another thread. Also cc'ed you there.

**Comment by liumomo315** [comment]
Hi Luke, sorry that I missed to read the "how to make review process smoother" and merged my commits locally and force pushed to origin. Will make sure not do that in the future. @lukecwik

**Comment by liumomo315** [comment]
Run Java PreCommit

File: sdks/python/apache_beam/portability/common_urns.py [hunk 4]
```diff
@@ -80,6 +80,10 @@ def PropertiesFromPayloadType(payload_type):
 session_windows = PropertiesFromPayloadType(
     standard_window_fns_pb2.SessionsPayload)
 
+# Custom window for test only.
+test_custom_windows = PropertiesFromPayloadType(
```
Comment (lukecwik) [inline sdks/python/apache_beam/portability/common_urns.py]: should not be defined here but locally within the test.

File: sdks/python/apache_beam/transforms/window_test.py (hunk 3 above)
Comment (lukecwik) [reply sdks/python/apache_beam/transforms/window_test.py]: It doesn't seem like there is a good way for custom user window fns to return a python dict or some other structure and you currently need to translate the data you want to store in the window fn into some proto. Instead of defining a test windowfn proto, use one of the well known protobuf types such as [Struct](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/struct) or [BytesValue](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/bytes-value) and embed the information within those objects yourself instead of defining your own within StandardWindowFns.

Comment (liumomo315) [reply sdks/python/apache_beam/transforms/window_test.py]: I made the custom window fn simpler in this version, by just overriding the assign() fn and the from_runner_api_parameter() fn. PTAL:) Thanks!

Comment (liumomo315) [reply sdks/python/apache_beam/transforms/window_test.py]: With that change, I also need to separate the windows runner api tests to a separate file.

File: sdks/python/apache_beam/transforms/window_test.py [hunk 5]
```diff
@@ -65,6 +66,30 @@ def process(self, element, window=core.DoFn.WindowParam):
...
+      return [IntervalWindow(0, 3)]
+    elif timestamp < 5:
+      return [IntervalWindow(3, 5)]
+    else:
+      return [IntervalWindow(5, timestamp)]
+
+  @urns.RunnerApiFn.register_urn(
+      common_urns.fixed_windows.urn,
+      standard_window_fns_pb2.FixedWindowsPayload)
+  def from_runner_api_parameter(fn_parameter, unused_context):
```
Comment (lukecwik) [inline sdks/python/apache_beam/transforms/window_test.py]: @robertwb did say that the default for to/from proto methods was to have it pickled so do we need to have this at all?

Comment (liumomo315) [reply sdks/python/apache_beam/transforms/window_test.py]: This custom window extends FixedWindows (to make the custom window implementation simple), so it needs to override the from_ fn. Otherwise the FixedWindows will be used in the pipeline.

Comment (liumomo315) [reply sdks/python/apache_beam/transforms/window_test.py]: Per offline discussion with Robert, the pylint error was a false alarm. Added "# pylint: disable=abstract-class-instantiated" to disable that check in the test. PTAL:) Thanks for your patient review, Luke!
